PermissionError: [WinError 32] The process cannot access the file because it is being used by another process

### 🐛 Describe the bug

The CI of Optuna integration, which uses Torch as a dependency, fails since the torch version was updated to v2.2.0.

Please check [here](https://github.com/optuna/optuna/actions/runs/7722855986/job/21051777682) for the error message.

With the version constraint `torch<2.2.0`, the same failure completely disappeared (Please check [here](https://github.com/optuna/optuna/actions/runs/7723123991/job/21052570860?pr=5217)).

### Versions

The environment is described in the Optuna CI.

https://github.com/optuna/optuna/actions/runs/7723123991/job/21052570860?pr=5217

Note that the same test will not fail on Ubuntu and Mac OS.
It fails only on Windows.

cc @ezyang @gchanan @zou3519 @kadeng @seemethere @malfet @osalpekar @atalman @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @peterjc123 @mszhanyi @skyline75489 @nbcsm @vladimir-aubrecht @iremyux @Blackhex @cristianPanaite

Minimal Repro:

```
(py311) C:\Users\runneruser\temp\pytorch>python test/distributed/test_device_mesh.py -k test_init_process_group
Fail to import hypothesis in common_utils, tests are not derandomized
C:\Users\runneruser\.conda\envs\py311\Lib\site-packages\torch\testing\_internal\common_distributed.py:553: ResourceWarning: unclosed file <_io.BufferedRandom name='C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\2\\tmpw4diowbl'>
  self.file_name = tempfile.NamedTemporaryFile(delete=False).name
ResourceWarning: Enable tracemalloc to get the object allocation traceback
Fail to import hypothesis in common_utils, tests are not derandomized
Fail to import hypothesis in common_utils, tests are not derandomized
Fail to import hypothesis in common_utils, tests are not derandomized
Fail to import hypothesis in common_utils, tests are not derandomized
[2024-01-31 22:08:32,424] torch.testing._internal.common_distributed: [ERROR] Caught exception:
[2024-01-31 22:08:32,424] torch.testing._internal.common_distributed: [ERROR] Traceback (most recent call last):
[2024-01-31 22:08:32,424] torch.testing._internal.common_distributed: [ERROR]   File "C:\Users\runneruser\.conda\envs\py311\Lib\site-packages\torch\testing\_internal\common_distributed.py", line 649, in run_test
[2024-01-31 22:08:32,424] torch.testing._internal.common_distributed: [ERROR]     getattr(self, test_name)()
[2024-01-31 22:08:32,424] torch.testing._internal.common_distributed: [ERROR]   File "C:\Users\runneruser\.conda\envs\py311\Lib\site-packages\torch\testing\_internal\common_distributed.py", line 535, in wrapper
[2024-01-31 22:08:32,424] torch.testing._internal.common_distributed: [ERROR]     fn()
[2024-01-31 22:08:32,424] torch.testing._internal.common_distributed: [ERROR]   File "C:\Users\runneruser\.conda\envs\py311\Lib\site-packages\torch\testing\_internal\common_utils.py", line 2652, in wrapper
[2024-01-31 22:08:32,424] torch.testing._internal.common_distributed: [ERROR]     method(*args, **kwargs)
[2024-01-31 22:08:32,424] torch.testing._internal.common_distributed: [ERROR]   File "C:\Users\runneruser\temp\pytorch\test\distributed\test_device_mesh.py", line 65, in test_init_process_group
[2024-01-31 22:08:32,424] torch.testing._internal.common_distributed: [ERROR]     DeviceMesh(device_type, mesh_tensor)
[2024-01-31 22:08:32,424] torch.testing._internal.common_distributed: [ERROR]   File "C:\Users\runneruser\.conda\envs\py311\Lib\site-packages\torch\distributed\device_mesh.py", line 202, in __init__
[2024-01-31 22:08:32,424] torch.testing._internal.common_distributed: [ERROR]     self._get_or_create_default_group()
[2024-01-31 22:08:32,424] torch.testing._internal.common_distributed: [ERROR]   File "C:\Users\runneruser\.conda\envs\py311\Lib\site-packages\torch\distributed\device_mesh.py", line 209, in _get_or_create_default_group
[2024-01-31 22:08:32,424] torch.testing._internal.common_distributed: [ERROR]     init_process_group()
[2024-01-31 22:08:32,424] torch.testing._internal.common_distributed: [ERROR]   File "C:\Users\runneruser\.conda\envs\py311\Lib\site-packages\torch\distributed\c10d_logger.py", line 86, in wrapper
[2024-01-31 22:08:32,424] torch.testing._internal.common_distributed: [ERROR]     func_return = func(*args, **kwargs)
[2024-01-31 22:08:32,424] torch.testing._internal.common_distributed: [ERROR]                   ^^^^^^^^^^^^^^^^^^^^^
[2024-01-31 22:08:32,424] torch.testing._internal.common_distributed: [ERROR]   File "C:\Users\runneruser\.conda\envs\py311\Lib\site-packages\torch\distributed\distributed_c10d.py", line 1177, in init_process_group
[2024-01-31 22:08:32,424] torch.testing._internal.common_distributed: [ERROR]     store, rank, world_size = next(rendezvous_iterator)
[2024-01-31 22:08:32,424] torch.testing._internal.common_distributed: [ERROR]                               ^^^^^^^^^^^^^^^^^^^^^^^^^
[2024-01-31 22:08:32,424] torch.testing._internal.common_distributed: [ERROR]   File "C:\Users\runneruser\.conda\envs\py311\Lib\site-packages\torch\distributed\rendezvous.py", line 246, in _env_rendezvous_handler
[2024-01-31 22:08:32,424] torch.testing._internal.common_distributed: [ERROR]     store = _create_c10d_store(master_addr, master_port, rank, world_size, timeout, use_libuv)
[2024-01-31 22:08:32,424] torch.testing._internal.common_distributed: [ERROR]             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2024-01-31 22:08:32,424] torch.testing._internal.common_distributed: [ERROR]   File "C:\Users\runneruser\.conda\envs\py311\Lib\site-packages\torch\distributed\rendezvous.py", line 174, in _create_c10d_store
[2024-01-31 22:08:32,424] torch.testing._internal.common_distributed: [ERROR]     return TCPStore(
[2024-01-31 22:08:32,424] torch.testing._internal.common_distributed: [ERROR]            ^^^^^^^^^
[2024-01-31 22:08:32,424] torch.testing._internal.common_distributed: [ERROR] torch.distributed.DistNetworkError: Unknown error
[2024-01-31 22:08:32,424] torch.testing._internal.common_distributed: [ERROR]
[2024-01-31 22:08:32,424] torch.testing._internal.common_distributed: [ERROR] To execute this test, run the following from the base repo dir:
[2024-01-31 22:08:32,424] torch.testing._internal.common_distributed: [ERROR]      python test\distributed\test_device_mesh.py -k test_init_process_group
[2024-01-31 22:08:32,424] torch.testing._internal.common_distributed: [ERROR]
[2024-01-31 22:08:32,424] torch.testing._internal.common_distributed: [ERROR] This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
[2024-01-31 22:08:32,424] torch.testing._internal.common_distributed: [ERROR]  exiting process 0 with exit code: 10
Process 0 terminated with exit code 10, terminating remaining processes.
E
======================================================================
ERROR: test_init_process_group (__main__.DeviceMeshTest.test_init_process_group)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\runneruser\.conda\envs\py311\Lib\site-packages\torch\testing\_internal\common_distributed.py", line 533, in wrapper
    self._join_processes(fn)
  File "C:\Users\runneruser\.conda\envs\py311\Lib\site-packages\torch\testing\_internal\common_distributed.py", line 752, in _join_processes
    self._check_return_codes(elapsed_time)
  File "C:\Users\runneruser\.conda\envs\py311\Lib\site-packages\torch\testing\_internal\common_distributed.py", line 802, in _check_return_codes
    raise RuntimeError(error)
RuntimeError: Process 0 exited with error code 10 and exception:
Traceback (most recent call last):
  File "C:\Users\runneruser\.conda\envs\py311\Lib\site-packages\torch\testing\_internal\common_distributed.py", line 649, in run_test
    getattr(self, test_name)()
  File "C:\Users\runneruser\.conda\envs\py311\Lib\site-packages\torch\testing\_internal\common_distributed.py", line 535, in wrapper
    fn()
  File "C:\Users\runneruser\.conda\envs\py311\Lib\site-packages\torch\testing\_internal\common_utils.py", line 2652, in wrapper
    method(*args, **kwargs)
  File "C:\Users\runneruser\temp\pytorch\test\distributed\test_device_mesh.py", line 65, in test_init_process_group
    DeviceMesh(device_type, mesh_tensor)
  File "C:\Users\runneruser\.conda\envs\py311\Lib\site-packages\torch\distributed\device_mesh.py", line 202, in __init__
    self._get_or_create_default_group()
  File "C:\Users\runneruser\.conda\envs\py311\Lib\site-packages\torch\distributed\device_mesh.py", line 209, in _get_or_create_default_group
    init_process_group()
  File "C:\Users\runneruser\.conda\envs\py311\Lib\site-packages\torch\distributed\c10d_logger.py", line 86, in wrapper
    func_return = func(*args, **kwargs)
                  ^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\runneruser\.conda\envs\py311\Lib\site-packages\torch\distributed\distributed_c10d.py", line 1177, in init_process_group
    store, rank, world_size = next(rendezvous_iterator)
                              ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\runneruser\.conda\envs\py311\Lib\site-packages\torch\distributed\rendezvous.py", line 246, in _env_rendezvous_handler
    store = _create_c10d_store(master_addr, master_port, rank, world_size, timeout, use_libuv)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\runneruser\.conda\envs\py311\Lib\site-packages\torch\distributed\rendezvous.py", line 174, in _create_c10d_store
    return TCPStore(
           ^^^^^^^^^
torch.distributed.DistNetworkError: Unknown error

To execute this test, run the following from the base repo dir:
     python test\distributed\test_device_mesh.py -k test_init_process_group

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0



----------------------------------------------------------------------
Ran 1 test in 3.729s

FAILED (errors=1)
```

--- a/torch/csrc/distributed/c10d/TCPStoreBackend.cpp
+++ b/torch/csrc/distributed/c10d/TCPStoreBackend.cpp
@@ -539,6 +539,7 @@ void TCPStoreMasterDaemon::run() {
       int rawSocket = socket.handle();
       sockets_.emplace_back(std::move(socket));
       tcputil::addPollfd(fds, rawSocket, POLLIN);
+      addMiscellaneousSocket(rawSocket);
     }
     queryFds(fds);
   }

